src/http.cc

```diff
@@ -1554,7 +1564,7 @@ bool
 HttpStateData::maybeMakeSpaceAvailable(bool doGrow)
 {
     // how much we are allowed to buffer
-    const int limitBuffer = (flags.headers_parsed ? Config.readAheadGap : Config.maxReplyHeaderSize);
+    const int limitBuffer = (flags.headers_parsed ? SQUID_TCP_SO_RCVBUF : Config.maxReplyHeaderSize);
```

Comment:
> Based on its actual usage, what is limitBuffer here? The maximum amount of data we can buffer in total? The maximum amount of data we can read assuming we have not read anything yet? Something else?

- While we are reading http headers reply this is the "the total number of bytes we can keep in memory, including the upcoming I/O bytes"
- While we are reading http response body bytes this is "the maximum bytes of data we can accept with one read from socket"

> As you have said, OS may allow Squid to read a lot more than SQUID_TCP_SO_RCVBUF/64KB. This change is definitely wrong from that point of view

well and yes and no:
  1. The SQUID_TCP_SO_RCVBUF if I am not wrong is exactly the receive buffer size in squid.
 2. I believe that the read buffer size should be configurable. We need one more configuration parameter in squid for this one. Definitely the readAHeadGap configuration parameter is used wrongly here
 3. If the connection to a remote http server is slow then most possible is that we do not need more than 10-16Kb for this buffer. In many cases an un-patched squid may use an input buffer  of 6MB (if readAHeadGap is 6Mb) to read 4-5Kb of data.  But if the connection to the remote server is very fast maybe a big buffer is needed here. But I suppose this is depends on other parameters (squid internals).
 4. I must note that the readAHeadGap by default in squid is 16Kb where the SQUID_TCP_SO_RCVBUF is 64Kb. Using SQUID_TCP_SO_RCVBUF instead of readAHeadGap improves squid for most of the squid installations.
 
> Would the correct version of this code then look like

My first impression is yes because we are emptying OS buffers and the OS  can accept more data from remote server. But is squid capable to forward enough fast these data to accept the next data? Or it is better to leave data to OS buffers so  the OS will not receive more data from remote server? 
I suppose it depends on application and probably is an system admin decision. A new network_read_buffer_size cfg parameter is a solution.

Reply:
> While we are reading http response body bytes this is "the maximum bytes of data we can accept with one read from socket"

That limitBuffer definition does not seem to match the existing code below:

```C++
if (limitBuffer < 0 || inBuf.length() >= (SBuf::size_type)limitBuffer) {
        // when buffer is at or over limit already
        debugs(11, 7, "will not read ...
        // Process next response from buffer
        processReply();
        return false;
    }
```

The code refuses to read more if inBuf already has limitBuffer bytes.

Reply:
When reading body: AFAICT, the current code (incorrectly) assumes that inBuf is where all read data is kept/accumulated for client use. In other words, the current code (incorrectly) ignores the fact that (based on your research) the read data is actually _moved_ into Store (MemObject) or Adaptation layer -- it is _that_ external layer that accumulates the data, not our inBuf. With that (incorrect) assumption in mind, the current code tries to limit inBuf.length() by Config.readAheadGap. In other words, limitBuffer is "the maximum amount of data we can buffer/accumulate in total".

The headers case is essentially the same except Config.readAheadGap becomes Config.maxReplyHeaderSize.

If the above summary is accurate, then I think we can leave the limitBuffer initialization code as is and define limitBuffer as "the maximum amount of virgin response data that Squid can buffer/accumulate in total". We also need to fix two primary bugs in this code (both can be attributed to the incorrect assumption discussed above):

1. The first `if` statement will continue to enforce limitBuffer but it should compute the total number of accumulated bytes instead of just using inBuf.length(). The condition is missing the component responsible for Store or Adaptation layer accumulation. I suspect some of the inner calcBufferSpaceToReserve() logic will need to be used here as well but that logic alone is not sufficient -- it does not account for Store accumulation AFAICT.

2. The read_size calculation may need to be adjusted as well. Just because we are willing to accumulate 200 MB worth of response data does not mean that we should increase inBuf to accommodate a single 200 MB read. I think the correct logic here would be something like this: We want to be able to read the whole OS socket buffer at a time (that is the maximum we could possible get if there were no gap limits) but we do not want to read any bytes that would violate the accumulation limit.

If you agree with the above logic, please implement it. For now, please use the `SQUID_TCP_SO_RCVBUF` constant for the "whole OS socket buffer" size. As we discussed, that constant is sometimes wrong. We will address that separately (possibly in this PR). Let's not try to solve _that_ problem for now.

Reply:
very-very tricky.
I made commit bd24e4c4e which probably does not have the form you are proposing here but I hope does the work. 

The HttpStateData::maybeMakeSpaceAvailable take its latest form after many adjusts.

I need to review the code when adaptation is used.

Please also check the XXX message over MemObject::readAheadPolicyCanRead method implementation.